datasets/pascalvoc.py

```diff
@@ -0,0 +1,238 @@
+    Torch metrics implementation is encapsulated.
+    (https://pytorch.org/vision/main/generated/torchvision.datasets.VOCDetection.html)
+    """
+
+    year_options = {'min': 2007, 'max': 2012}
+
+    # For years 2008 and above, supported classes are common for all:
+    voc_labels = ('aeroplane', 'bicycle', 'bird', 'boat', 'bottle', 'bus',
+                  'car', 'cat', 'chair', 'cow', 'diningtable', 'dog',
+                  'horse', 'motorbike', 'person', 'pottedplant', 'sheep', 'sofa',
+                  'train', 'tvmonitor')
+
+    voc_label_to_id_map = {k: v + 1 for v, k in enumerate(voc_labels)}
+    voc_id_to_label_map = {v: k for k, v in voc_label_to_id_map.items()}
+
+    def __init__(self, root, d_type, img_size, year_opt, augment_data,
+                 transform, download=False):
+
+        self.data_root = os.path.join(root, self.__class__.__name__)
+        self.img_size = img_size
+        self.augment_data = augment_data
+        self.transform = transform
+
+        self.__check_for_download(year_opt, d_type, download)
+        self.__prepare_dataset(d_type)
+
+    def __check_for_download(self, year_opt, d_type, download):
+        self.year_list = []
+        self.download = {}
+
+        year_opt_list = year_opt.split('_')
+        for year_str in year_opt_list:
+
+            if int(year_str) >= self.year_options['min'] and \
+               int(year_str) <= self.year_options['max']:
+
+                self.year_list.append(year_str)
+                self.download[year_str] = download and \
+                    (not self.__check_data_path_exists(year_str, d_type))
+            else:
+                print(f'Warning: {year_str} is not a valid year for Pascal VOC!')
+
+    def __check_data_path_exists(self, year_str, d_type):
+        return os.path.exists(os.path.join(self.data_root, 'VOCdevkit',
+                                           'VOC' + year_str, 'ImageSets',
+                                           'Main', d_type + '.txt'))
+
+    def __prepare_dataset(self, dtype):
+
+        self.dataset = None
+
+        for year_str in self.year_list:
+            dataset_temp = \
+                torchvision.datasets.VOCDetection(root=self.data_root, year=year_str,
+                                                  image_set=dtype,
+                                                  download=self.download[year_str],
+                                                  transforms=self.__transforms_func)
+
+            if self.dataset:
+                self.dataset = torch.utils.data.ConcatDataset([self.dataset, dataset_temp])
+            else:
+                self.dataset = dataset_temp
+
+    def __transforms_func(self, image, target):
+        """
+        Torch vision VOCDetection dataset's __get_item__ returns: (image, target)
+        where target is a dictionary of the XML tree.
+        This transform function applies transformations on:
+            * images: Normalized, resized, ai8xized
```

Comment:
the image should not be ai8xized at this level. This transformation is applied at the very end; in __get_item method.

Reply:
Correct, the comments are revised in the last commit